terraform/eval_read_data_apply.go

```diff
@@ -0,0 +1,98 @@
+package terraform
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/plans"
+	"github.com/hashicorp/terraform/states"
+	"github.com/hashicorp/terraform/tfdiags"
+)
+
+// EvalReadDataApply is an EvalNode implementation that deals with the main part
+// of the data resource lifecycle: either actually reading from the data source
+// or generating a plan to do so.
+type EvalReadDataApply struct {
+	evalReadData
+}
+
+func (n *EvalReadDataApply) Eval(ctx EvalContext) (interface{}, error) {
+	absAddr := n.Addr.Absolute(ctx.Path())
+
+	var diags tfdiags.Diagnostics
+
+	var planned *plans.ResourceInstanceChange
+	if n.Planned != nil {
+		planned = *n.Planned
+	}
+
+	if n.ProviderSchema == nil || *n.ProviderSchema == nil {
+		return nil, fmt.Errorf("provider schema not available for %s", n.Addr)
+	}
+
+	if planned != nil && !(planned.Action == plans.Read || planned.Action == plans.Update) {
+		// If any other action gets in here then that's always a bug; this
+		// EvalNode only deals with reading.
+		return nil, fmt.Errorf(
+			"invalid action %s for %s: only Read or Update is supported (this is a bug in Terraform; please report it!)",
+			planned.Action, absAddr,
+		)
+	}
+
+	if err := ctx.Hook(func(h Hook) (HookAction, error) {
+		return h.PreApply(absAddr, states.CurrentGen, planned.Action, planned.Before, planned.After)
+	}); err != nil {
+		return nil, err
+	}
+
+	// we have a change and it is complete, which means we read the data
+	// source during plan and only need to store it in state.
+	if planned.Action == plans.Update {
```

Comment:
Overloading the meaning of `Update` here makes me nervous. :confused: I can't identify a specific reason why right now, but this meaning of "update" seems different enough to at least cause some confusion in future maintenance.

Part of me wants to suggest adding a new verb here like "Commit" to represent just writing something to the state without any other side-effect, but at the same time that's also a pretty big change in our data model that I wouldn't want to rush into without some more design thought first. :confounded: 

Given that we're planning to merge plan and refresh into a single step later anyway, and at that point we should no longer need to distinguish these two cases (because the updated data will be in the "prior state" already), is there room for a temporary slightly-more-tricky solution here, such as checking whether the "before" object in the plan contains unknown values? I'm not pretending that this is any easier to understand than overloading `Update`, but I'm hoping it'd be a more _localized_ temporary hack that would be easier to mark with cautionary comments in one spot and easier to find and remove later when we are ready to do the full refresh/plan merger.... :thinking: 

(By "localized" here I mean that it would hopefully be less visible to other subsystems like our plan renderer, JSON plan output, etc, and just confined to the logic in this package. :crossed_fingers: I'm particularly worried about it being visible in the JSON plan output, because Sentinel policies can then create an additional compatibility constraint for us to clean it up later.)

Reply:
I went around and around on this a bit, and for some reason the idea that "updating" a data source meant updating its state is what fell out as the simplest use case. I did leave that open for changing however; but if you think about a data source as a resource which only has a plan, the apply step simply becomes saving the planned value.

I'm also leaning towards keeping the concepts of data resources and manages resources as closely aligned as possible (i.e. `ReadDataSource` is just a special form of `PlanResourceChange` that returns all values). This is another reason I like this idea of data source diffs showing up as planned changes. Having data sources be silently read only sometimes, and have planned changes at other times, definitely leads to some of the confusion around their behavior. 

Sentinel making use of this without our intention to support it is a good point though, so I think we just need to decide if this is a good method to stick with. If the goal is to eventually have data source changes show up in the plan (which there is a lot of demand for, though it could be done in other ways), and `Update` is an acceptable verb for the situation, I think this is a good step forward. If we're unsure about the final use case, we should change the verb to something new, and hide the change from the json output for now.

Reply:
The refactoring here actually left the data source plan handling in the perfect place to handle this transparently. The `change.After` value can now be used directly to determine whether apply needs to read the data source or not. 

This alleviates us from designing any new plan features for data sources at the moment.